source/extensions/dynamic_modules/abi.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This is a pure single C header file that defines the ABI of the dynamic module
+// used by Envoy.
+//
+// This must not contain any dependencies besides standard library since it is not only used by
+// Envoy itself but also by dynamic module SDKs written in non-C++ languages.
+//
+// Currently, compatibility is only guaranteed by an exact version match between the Envoy
+// codebase and the dynamic module SDKs. In the future, after the ABI is stabilized, we will revisit
+// this restriction and hopefully provide a wider compatibility guarantee. Until then, we use the
+// simple symbol prefix "envoy_dynamic_module_" without any versioning at the symbol level.
```

Comment:
I think this is going to lead to hard to diagnose issues. What is the plan for how to block loading modules with the wrong version? Should we compile in the build SHA somehow and check that they match on each side?

Reply:
so I was thinking about the SHA matching suggestion you gave before but i realized that it's going to be really hard as the end shared object file might be compiled outside the Envoy source tree, hence they might have a different hash. One idea would be to directly encode `some_hash_function(abi.h)` into each SDK code and the Envoy side `dynamic_modules.h` with some failure check. Let me create some prototype. I would appreciate it if you have another idea (Forcing users to clone Envoy and develop in-tree works, but sounds not great for users to me).

Reply:
I would have to think about it more. I'm not sure the best solution. One option I suppose would be to export the ABI file(s) during build time so that the SHA can be embedded in there and then passed through during registration.

Reply:
I pushed the change that implements my idea. That is basically hard code the `sha256(abi.h)` into `abi_version.h`, and the dynamic modules must returns exactly the same hash during the program init function. Let me know what you feel about it

Reply:
I like the way you implemented this. As I think most will build their dynamic modules outside of the Envoy codebase and it would be complex to require development in-tree (in some form). For the SDKs, they will work with a specific `sha256(abi.h)` and I think that is 👍. As a platform engineer I'd write some plugin integration logic with the specific Envoy versions I'd target. I think this allows me to do that.

Reply:
I tried to figure out how to implement the Matt's idea on embedding SHA during the build time into files, but I couldn't find a realistic way to do so - Bazel in general seems not to allow edit the source files

Reply:
I think that simple is preferable in this case. We can add some details to the development README which have minimal steps to update with one of those steps being updating `static const char* version`. I don't mind the manual update as this then needs to be part of the PR review process. Also, it means that there is no magic happening in the Bazel, which I think is a welcomed thing.